pace-util/pace/util/report.py

```diff
@@ -0,0 +1,114 @@
+import copy
+import json
+from datetime import datetime
+from typing import Any, Dict, List
+
+import numpy as np
+
+from pace.util.mpi import MPI
+
+
+def set_experiment_info(
+    experiment_name: str, time_step: int, backend: str, git_hash: str
+) -> Dict[str, Any]:
+    experiment: Dict[str, Any] = {}
+    now = datetime.now()
+    dt_string = now.strftime("%d/%m/%Y %H:%M:%S")
+    experiment["setup"] = {}
+    experiment["setup"]["timestamp"] = dt_string
+    experiment["setup"]["dataset"] = experiment_name
+    experiment["setup"]["timesteps"] = time_step
+    experiment["setup"]["hash"] = git_hash
+    experiment["setup"]["version"] = "python/" + backend
+    experiment["setup"]["format_version"] = 3
+    experiment["times"] = {}
+    return experiment
+
+
+def collect_keys_from_data(times_per_step: List[Dict[str, float]]) -> List[str]:
+    """Collects all the keys in the list of dics and returns a sorted version"""
+    keys = set()
+    for data_point in times_per_step:
+        for k, _ in data_point.items():
+            keys.add(k)
+    sorted_keys = list(keys)
+    sorted_keys.sort()
+    return sorted_keys
+
+
+def gather_timing_data(
+    times_per_step: List[Dict[str, float]],
+    results: Dict[str, Any],
+    comm: MPI.Comm,
+    root: int = 0,
+) -> Dict[str, Any]:
+    """returns an updated version of  the results dictionary owned
+    by the root node to hold data on the substeps as well as the main loop timers"""
+    is_root = comm.Get_rank() == root
+    keys = collect_keys_from_data(times_per_step)
+    data: List[float] = []
+    for timer_name in keys:
+        data.clear()
+        for data_point in times_per_step:
+            if timer_name in data_point:
+                data.append(data_point[timer_name])
+
+        sendbuf = np.array(data)
+        recvbuf = None
+        if is_root:
+            recvbuf = np.array([data] * comm.Get_size())
+        comm.Gather(sendbuf, recvbuf, root=0)
+        if is_root:
+            results["times"][timer_name]["times"] = copy.deepcopy(recvbuf.tolist())
+    return results
+
+
+def write_global_timings(experiment: Dict[str, Any]) -> None:
+    now = datetime.now()
+    filename = now.strftime("%Y-%m-%d-%H-%M-%S")
+    with open(filename + ".json", "w") as outfile:
+        json.dump(experiment, outfile, sort_keys=True, indent=4)
```

Comment:
Other than the filename used, is there anything that makes this function specific to writing global timings? It seems like `write_to_timestamped_json` might be more descriptive.

Reply:
Changed to `write_to_timestamped_json`